File: leetcode-go/Q051-Q100/Q094_Binary_Tree_Inorder_Traversal/Q094_Binary_Tree_Inorder_Traversal.go
package Q094_Binary_Tree_Inorder_Traversal

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func inorderTraversal(root *TreeNode) []int {
	ans := []int{}
	dfs(&ans, root)
	return ans
}

func dfs(ans *[]int, root *TreeNode) {
	if root == nil {
		return
	}
	dfs(ans, root.Left)
	*ans = append(*ans, root.Val)
	dfs(ans, root.Right)
}

File: leetcode-java/src/vmware/tree_traversal/Q094_Binary_Tree_Inorder_Traversal.java
package vmware.tree_traversal;

import java.util.*;

public class Q094_Binary_Tree_Inorder_Traversal {

    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) { val = x; }
    }

    public List<Integer> inorderTraversal(TreeNode root) {
        Stack<TreeNode> stack = new Stack<>();
        while (root != null) {
            stack.push(root);
            root = root.left;
        }
        List<Integer> result = new LinkedList<>();
        while (!stack.isEmpty()) {
            TreeNode curr = stack.pop();
            result.add(curr.val);
            curr = curr.right;
            while (curr != null) {
                stack.push(curr);
                curr = curr.left;
            }
        }
        return result;
    }

}
